Add native-image metadata for the different ChannelHandler implementations

gradinac opened the pull request:
Motivation:

Due to a specific reflection usage pattern in `io.netty5.channel.ChannelHandlerMask#mask0`, the agent cannot automatically infer the necessary conditional metadata. Therefor, we must manually add such metadata for subclasses of the `ChannelHandler`.

See also https://github.com/netty/netty/issues/12725#issuecomment-1225694049

Modification:

This PR adds the necessary metadata for handlers used in the Netty 5 testsuite.

Result:

Should fix [#12725](https://github.com/netty/netty/issues/12725)

Note that for user-defined `ChannelHandler` implementations, users will still need to supply necessary metadata, either manually or by using the tracing agent

gradinac commented:
cc @violetagg and thank you for helping with testing and cleaning up of the generated metadata :)

normanmaurer commented:
@gradinac its really "sad" that we need to always manually update these whenever we add more stuff :( Is there no other way to make this work ?

gradinac commented:
I agree completely, manual metadata maintenance = bad

Right now, we don't have anything on the native-image/native-image-agent side that would help with this particular issue :/ I can push a python script to automate generating these from the collected metadata (and convert the existing bash script that collects metadata into a python script)

If the above is okay, I can create a follow up PR with the above python script

chrisvest commented:
It would be good to have a python script. We need to have a build step that checks if this metadata is up to date, so breakage can't sneak in.

pderop commented:
@chrisvest ,

I did not have time to fully dive into all the details, but in Reactor-Netty, we have a Junit test which scans all handlers using reflection, and the test will fail the build in case one handler is missing from the reflect-config.json.

See https://github.com/reactor/reactor-netty/pull/2432

More specifically, see the [junit test](https://github.com/reactor/reactor-netty/blob/main/reactor-netty-http/src/test/java/reactor/netty/NativeConfigTest.java) which performs the checking.

gradinac commented:
@pderop using the Reflections library in a JUnit test to find all implementations of the `ChannelHandler` would be nice, as we could both generate and verify this metadata, however, I'm not sure how it would best be implemented in Netty
One requirement of that approach would be to have all Netty modules on the classpath for that test

violetagg commented:
> @pderop using the Reflections library in a JUnit test to find all implementations of the `ChannelHandler` would be nice, as we could both generate and verify this metadata, however, I'm not sure how it would best be implemented in Netty
> One requirement of that approach would be to have all Netty modules on the classpath for that test

I think that `netty-all` has all dependencies

normanmaurer commented on common/src/main/resources/META-INF/native-image/handlers/reflect-config.json:
```diff
@@ -0,0 +1,552 @@
+[
```
Also one thing I was wonder was if we should really put this in common ? Shouldn't we have the config with the right entries in each artefact that has such a handler ?

gradinac replied:
That's a good point - the reason why I put them under the common directory is because these would be available to every project, so we would only have to maintain it in one place. During the image build, only handlers that are reachable (both available on the classpath and used in the application) would be included

I'm looking into the reflections approach for automatically generating this metadata now - if this approach ends up working, I could split the generated metadata on a per-project basis

gradinac commented:
I've tried the `org.reflections` based approach on this PR - in essence, it adds the `ChannelHandlerMetadataUtil` class that finds all subclasses of `ChannelHandler` in the given packages, loads up existing metadata of the target module and then compares whether there should be any changes to the metadata.
Each Netty module that subtypes `ChannelHandler` has a small unit test that specifies where it's metadata file lives and which packages should be searched in that module.
I've currently added this test for the `transport` and `codec` modules, and used it to automatically generate metadata.

The benefits of this approach is that it would run as a standard unit test every time there's a change and point the exact different in metadata (if any), as well as that it would generate metadata on a per-module basis, rather than storing it as a monolithic file inside `netty5-common`

The downsides are that we have to add a bit of boilerplate code to each project (3 `pom.xml` dependencies - org.reflections, gson and a test jar dependency on `netty5-transport`) - if there's a smarter way to do this, I can give a go at implementing it :)

Would this approach be okay? If yes, I can add the necessary bits to the rest of the Netty modules and polish up the PR

gradinac commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+                System.err.println();
+            }
+
+            System.err.println("Expected metadata file contents:\n");
```
Maybe we could automatically update the metadata file here, rather than printing a message and letting the users do it

normanmaurer replied:
I think its fine for now

gradinac commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+        Set<Class<? extends ChannelHandler>> classes = reflections.getSubTypesOf(ChannelHandler.class);
+        classes = classes.stream()
+                .filter(ChannelHandlerMetadataUtil::isTestClass)
+                .filter(clazz -> Stream.of(packageNames).anyMatch(name -> clazz.getName().startsWith(name)))
```
For some reason, `org.reflections` still finds classes in packages other than the ones we specified, so we do a little bit of manual cleanup here

normanmaurer replied:
nit: why not just chain the `stream...` directly ? this way you not need re-assign `classes`

gradinac replied:
That's a good point, with that change we can completely remove the `classes` variable

gradinac commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+        if (parts.length > 0) {
+            URL classFile = clazz.getResource(parts[parts.length - 1] + ".class");
+            if (classFile != null) {
+                return !classFile.toString().contains(File.separator + "test-classes" + File.separator);
```
This one bugged my mind a bit - is there a smarter way to find out if a class originates from the test sources? It would be nice to have all of the test classes filtered automatically, and this seems to work, but it also seems a bit ugly

gradinac commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+                .map(type -> new HandlerMetadata(type.getName(), new Condition(type.getName()), true))
+                .collect(Collectors.toSet());
+
+        Path existingMetadataPath = new File(projectRelativeResourcePath).toPath().toAbsolutePath();
```
Maven seems to always set the current directory to the module it's currently building/testing - this lets us look up the `reflect-config.json` file directly from the module sources by using a relative path here

gradinac commented on codec/pom.xml:
```diff
@@ -100,6 +100,22 @@
+      <groupId>com.google.code.gson</groupId>
+      <artifactId>gson</artifactId>
+    </dependency>
+    <dependency>
```
Ideally we don't want to re-write the metadata generation for every test, but we also don't want to bloat up the non-test sources with test specific code. This creates a test-only dependency to the test-jar of `netty5-transport` where the metadata-generation logic lives

gradinac commented:
It seems like the "is this class a test class" test is failing on Windows - most likely because I messed up the path checking logic. I'll look into it

chrisvest commented:
You'll also need to rebase to avoid the build failing on some certificate validation. Some test certificates expired and have been replaced.

gradinac commented:
@chrisvest Thank you for the heads up! :)
I've rebased the PR to the latest master and I've fixed the Windows metadata generation issue in [c3134de](https://github.com/netty/netty/pull/12738/commits/c3134de2d63e91330a65a1bb7a9054e927e9bdd6).

I've also added the logic for automatic metadata generation to remaining Netty modules that needed it

normanmaurer commented:
@gradinac I am correct to assume we would also need this for 4.1 ?

normanmaurer commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+
+            System.err.println("Expected metadata file contents:\n");
+            System.err.println(getMetadataJsonString(handlerMetadata));
+            Assertions.fail();
```
Why not add what you pass to `System.err.println()` to the fail method ?

gradinac replied:
That's a good point, I'll move all `System.err.println` messages to the `fail` method instead

normanmaurer commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+                    "This metadata was not found under " + existingMetadataPath);
+            System.err.println("Please create this file with the following content: ");
+            System.err.println(getMetadataJsonString(handlerMetadata));
+            Assertions.fail();
```
Why not add what you pass to `System.err.println()` to the fail method ?

normanmaurer commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+        return gson.toJson(metadataList, HANDLER_METADATA_LIST_TYPE);
+    }
+
+    private static class Condition {
```
```suggestion
    private static final class Condition {
```

normanmaurer commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+            this.typeReachable = typeReachable;
+        }
+
+        String typeReachable;
```
```suggestion
        final String typeReachable;
```

normanmaurer commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    private static class HandlerMetadata {
+        String name;
```
```suggestion
        final String name;
```

normanmaurer commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+    private static class HandlerMetadata {
+        String name;
+
+        Condition condition;
```
```suggestion
        final Condition condition;
```

normanmaurer commented on transport/src/test/java/io/netty5/nativeimage/ChannelHandlerMetadataUtil.java:
```diff
@@ -0,0 +1,217 @@
+
+        Condition condition;
+
+        boolean queryAllPublicMethods;
```
```suggestion
        final boolean queryAllPublicMethods;
```

gradinac commented:
> @gradinac I am correct to assume we would also need this for 4.1 ?

This is correct - I didn't want to create a Netty 4.1 PR yet though as I wasn't sure if I had taken the right approach for generating this metadata

gradinac commented:
I've pushed the changes that should address the review - in the meantime, I've also prepared a Netty 4 PR and gathered the required metadata. Once that metadata is tested, I'll create a Netty 4 PR as well

normanmaurer commented:
Thanks a lot! Great work

gradinac commented:
Thank you very much for looking into this! :) I'll have a PR for Netty 4.1 ready soon with the same approach
